Added offsets.

jcarrier0026 opened the pull request:
I added the ability to add offsets to every frame in an animation.

jcarrier0026 committed:
Added offsets.

I added the ability to add offsets to every frame in an animation.

mrmarkwell commented on animation.h:
```diff
@@ -3,13 +3,24 @@
+  kDefault,
+  kRunRight,
+  kRunLeft,
+  kSSTransformation,
```
Even for acronyms (SS for super saiyan), use CamelCase

kSsTransformation

https://google.github.io/styleguide/cppguide.html#Constant_Names

mrmarkwell commented on animation.h:
```diff
@@ -3,13 +3,24 @@
   int time_between_frames_ms;
 
+  // Offsets for each frame.
+  std::vector<Location> offsets;
```
This is similar to having two maps for animation information, and the solution is the same. Associated data should be grouped together: make a struct that holds an SDL_Rect and a Location so that there isn't an implicit association with the two separate vectors.

mrmarkwell commented on animation_data.cpp:
```diff
@@ -20,25 +25,76 @@ std::vector<SDL_Rect> BuildAnimationRects(SDL_Rect start,
+                        {-29, -55},
+                        {-29, -55},
+                        {-29, -55}},
+            .image_rects = BuildAnimationRects(kGokuSSTransformationStart,
```
make a version of BuildAnimationRects that takes the offsets as an argument instead of the number of frames:

For illustration I'm calling the struct with the rect and offsets an "AnimationFrame":

```
std::vector<AnimationFrame> BuildAnimationRects(SDL_Rect start, std::vector<Location> offsets) { ... }
```

Use this version when there are offsets, and the other one will be for when all the offsets are 0, 0.

mrmarkwell commented on graphics.cpp:
```diff
@@ -50,10 +50,12 @@ void Graphics::AddSprite(const Sprite& sprite, const SDL_Rect& destination,
 
 void Graphics::AddSprite(const Sprite& sprite, Location location, Perf* perf) {
   SDL_Rect source = sprite.GetSourceLocation();
```
Make this "GetSourceAnimationFrame", and return the struct with the offset and rect

mrmarkwell commented on sprite.h:
```diff
@@ -28,6 +29,8 @@ class Sprite {
   // Returns the scale.
   virtual float GetScale() const { return scale_; }
 
+  virtual Location GetOffsets() const { return {0, 0}; };
```
Remove

mrmarkwell commented on animation_data.cpp:
```diff
@@ -20,25 +25,76 @@ std::vector<SDL_Rect> BuildAnimationRects(SDL_Rect start,
            {.type = AnimationType::kRunRight,
             .time_between_frames_ms = 100,
+            .offsets = NoOffsets(/*number_of_frames*/ 2),
             .image_rects =
```
let's rename this to animation_frames (or whatever you decide to call the struct)

mrmarkwell commented on animation_data.cpp:
```diff
@@ -20,25 +25,76 @@ std::vector<SDL_Rect> BuildAnimationRects(SDL_Rect start,
+
+// Takes the number of frames and sets the offset for each frame of the
+// animation to zero.
+std::vector<Location> NoOffsets(int number_of_frames) {
+  std::vector<Location> result;
+  for (int i = 0; i < number_of_frames; i++) {
+    result.push_back({0, 0});
+  }
```
Remove

mrmarkwell commented on animated_sprite.h:
```diff
@@ -36,6 +36,8 @@ class AnimatedSprite : public Sprite {
   // Returns the current source location for the animated sprite.
   SDL_Rect GetSourceLocation() const override;
 
+  Location GetOffsets() const override;
```
remove

mrmarkwell commented on animated_sprite.h:
```diff
@@ -36,6 +36,8 @@ class AnimatedSprite : public Sprite {
   // Returns the current source location for the animated sprite.
   SDL_Rect GetSourceLocation() const override;
```
GetSourceLocation is a bad name now, since "Location" is a different concept than "SDL_Rect".

Let's make this return an AnimationFrame anyways

The base class Sprite can just return an animation frame with offset 0, 0

mrmarkwell commented on animated_sprite.cpp:
```diff
@@ -24,6 +24,10 @@ SDL_Rect AnimatedSprite::GetSourceLocation() const {
   return animations_.at(current_animation_type_).image_rects[frame_index_];
 }
 
+Location AnimatedSprite::GetOffsets() const {
```
Remove

mrmarkwell reviewed (changes requested):
Great work! I'd like you to group the offsets with the image rect using a struct. Otherwise it looks good.

jcarrier0026 committed:
Fixed requested changes.

Added an AnimationFrame struct to hold the source rect and the offset together.

jcarrier0026 committed:
Forgot to add a file to last commit.

Here is the missing file for the last commit.

mrmarkwell committed:
Merge branch 'develop' into feature/offset

mrmarkwell commented on animation_frame.h:
```diff
@@ -0,0 +1,13 @@
+  SDL_Rect source;
+  Location offset;
+
+  AnimationFrame() : source({}), offset(0, 0) {}
+  AnimationFrame(SDL_Rect rect, Location location)
+      : source(rect), offset(location) {}
```
I'm not convinced these are needed, but not a big deal to have them. Struct members are public by default, so I don't know that it adds much value to have a constructor.

mrmarkwell reviewed (approved):
Looks great, feel free to merge.

jcarrier0026 committed:
Merged Perf changes with the Offsets changes.

Got the code with both features working.
